### DISCUSSION TITLE
unable to control the ur10e arm using moveit2

### DISCUSSION THREAD

**Discussion by Amina-axon** [UniversalRobots/Universal_Robots_ROS2_Driver#559] [Q&A]
i am trying currently to move my ur10e using moveit2. i am working with ros2 Humble on a ubuntu 22.04 virtual machine. it but it is causing some issues so far.
this is how i try to control it :
i have the external control installed on my TP with my pc IP address.

1. i run the following command to launch the driver:

_ros2 launch ur_robot_driver ur_control.launch.py ur_type:=ur10e robot_ip:=192.168.102.210 launch_rviz:=false initial_joint_controller:=joint_trajectory_controller_

2. i press on the play buttom for the external control program on my TP

Then i get the following output at the end : 
**[ur_ros2_control_node-1] [INFO] [1670255474.584292674] [UR_Client_Library]: Robot requested program

[ur_ros2_control_node-1] [INFO] [1670255474.584354385] [UR_Client_Library]: Sent program to robot

[ur_ros2_control_node-1] [INFO] [1670255474.660009241] [UR_Client_Library]: Robot connected to reverse interface. Ready to receive control commands.

[ur_ros2_control_node-1] [WARN] [1670255474.661654264] [controller_manager]: 'start_controllers' field is deprecated, use 'activate_controllers' field instead!**

3. i launch moveit2 using the following command:
_ros2 launch ur_moveit_config ur_moveit.launch.py ur_type:=ur10e robot_ip:=192.168.102.209 launch_rviz:=true initial_joint_controller:=joint_trajectory_controller_

after that i can see on Rviz that it shows the current position of the robot . when i try to plan and execute a motion it fails with this error:

**[move_group-1] [INFO] [1670255494.899591323] [moveit_ros.trajectory_execution_manager]: Starting trajectory execution ...

[move_group-1] [INFO] [1670255494.900450181] [moveit.plugins.moveit_simple_controller_manager]: Returned 2 controllers in list

[move_group-1] [INFO] [1670255494.900726877] [moveit.plugins.moveit_simple_controller_manager]: Returned 2 controllers in list

[move_group-1] [ERROR] [1670255494.901352459] [moveit.simple_controller_manager.follow_joint_trajectory_controller_handle]: Action client not connected to action server: joint_trajectory_controller/follow_joint_trajectory

[move_group-1] [ERROR] [1670255494.901800931] [moveit_ros.trajectory_execution_manager]: Failed to send trajectory part 1 of 1 to controller joint_trajectory_controller

[move_group-1] [INFO] [1670255494.901981690] [moveit_ros.trajectory_execution_manager]: Completed trajectory execution with status ABORTED ...

[move_group-1] [INFO] [1670255494.904622340] [moveit_move_group_default_capabilities.move_action_capability]: Solution found but controller failed during execution

[rviz2-2] [INFO] [1670255494.908250297] [move_group_interface]: Plan and Execute request aborted

[rviz2-2] [ERROR] [1670255494.979549461] [move_group_interface]: MoveGroupInterface::move() failed or timeout reached**

 and i got this on the terminal where i have the driver launched:

**[ur_ros2_control_node-1] [INFO] [1670255486.079104794] [UR_Client_Library]: Connection to reverse interface dropped.

[ur_ros2_control_node-1] [INFO] [1670255486.102143386] [UR_Client_Library]: Robot connected to reverse interface. Ready to receive control commands.

[ur_ros2_control_node-1] [WARN] [1670255486.105222094] [controller_manager]: 'start_controllers' field is deprecated, use 'activate_controllers' field instead!

[ur_ros2_control_node-1] [WARN] [1670255499.925256433] [UR_Client_Library]: Connection attempt on port 50001 while maximum number of clients (1) is already connected. Closing connection.

[ur_ros2_control_node-1] [INFO] [1670255499.930122049] [UR_Client_Library]: Connection to reverse interface dropped.**

in some occasion my motion execution was successful but on other occasion it shows the problem above.

is there any suggestion of what i might be doing wrong or if there is another approach to follow?

**Comment by fmauch** [comment]
> `ros2 launch ur_robot_driver ur_control.launch.py ur_type:=ur10e robot_ip:=192.168.102.210 launch_rviz:=false initial_joint_controller:=joint_trajectory_controller`

the `initial_joint_controller:=joint_trajectory_controller` should only be used with fake_hardware. Since you are trying to connect to a real robot, that should not be used.

> `ros2 launch ur_moveit_config ur_moveit.launch.py ur_type:=ur10e robot_ip:=192.168.102.209 launch_rviz:=true initial_joint_controller:=joint_trajectory_controller`

For the moveit launchfile the `initial_joint_controller` argument is not supposed to do anything. In case of using fake_hw, one should pass `use_fake_hardware:=true` as explained in the [docs](https://docs.ros.org/en/ros2_packages/rolling/api/ur_robot_driver/usage.html#using-moveit)

Looks like you almost succeeded, just don't pass the `initial_joint_controller:=joint_trajectory_controller` to any of the launchfiles and you should be good to go.

**Reply by Amina-axon** [reply]
Thanks @fmauch ! but not passing the _initial_joint_controller_ didn't fix the problem.
however , going through this issue https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/issues/507 and  adding the line **ur_driver_->setKeepaliveCount(5);** to my hardware_interface.cpp fixed the issue and now my robot is moving without any problem.
